services/webnn/ort/graph_builder_ort.cc

```diff
@@ -952,6 +986,123 @@ GraphBuilderOrt::AddExpandOperation(const mojom::Expand& expand) {
   return base::ok();
 }
 
+[[nodiscard]] base::expected<void, mojom::ErrorPtr>
+GraphBuilderOrt::AddDequantizeLinearOperation(
+    const mojom::DequantizeLinear& dequantize_linear) {
+  const std::string node_name =
+      GenerateNextOperationName(dequantize_linear.label);
+  std::string input_name =
+      GetOperandNameById(dequantize_linear.input_operand_id);
+  std::string scale_name =
+      GetOperandNameById(dequantize_linear.scale_operand_id);
+  std::string zero_point_name =
+      GetOperandNameById(dequantize_linear.zero_point_operand_id);
+  std::string output_name =
+      GetOperandNameById(dequantize_linear.output_operand_id);
+
+  const OperandDescriptor& input_descriptor =
+      GetOperand(dequantize_linear.input_operand_id).descriptor;
+  std::vector<uint32_t> input_shape = input_descriptor.shape();
+
+  const OperandDescriptor& scale_descriptor =
+      GetOperand(dequantize_linear.scale_operand_id).descriptor;
+  std::vector<uint32_t> scale_shape = scale_descriptor.shape();
+
+  int64_t axis = 1;
+  int64_t block_size = 0;
+  bool need_transpose = false;
+
+  // https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/frontend/src/op/dequantize_linear.cpp#L220
+  if (scale_shape.size() > 2) {
+    return NewNotSupportedError(
+        "OpenVINO dequantizeLinear cannot operate with more than 2D scales");
+  }
+
+  if (scale_shape.empty()) {
+    // For per-tensor/layer dequantization the scale is a scalar.
+    axis = 0;
+  } else if (scale_shape.size() == 1) {
+    bool is_valid = false;
+    // for per per-axis dequantization it is a 1-D Tensor
```

Comment:
Re WebNN dequantizeLinear definition, the 1D scale only applies to the last dimension, for example input shape [3, 4, 5], scale shape [5]. For other dimensions, WebNN scale operand should have the all size one shape except input size is equal to scale size along axis, for example input shape [3, 4, 5], scale shape [1, 4, 1]. BTW, [Dwayne is proposing to have input and scale has the same rank](https://github.com/webmachinelearning/webnn/pull/805#discussion_r1919498405), so for the first example, the input shape [3, 4, 5], scale shape should be [1, 1, 5].

Reply:
> Re WebNN dequantizeLinear definition, the 1D scale only applies to the last dimension, for example input shape [3, 4, 5], scale shape [5]. For other dimensions, WebNN scale operand should have the all size one shape except input size is equal to scale size along axis, for example input shape [3, 4, 5], scale shape [1, 4, 1].

Agree. I will modify the code.

> so for the first example, the input shape [3, 4, 5], scale shape should be [1, 1, 5].

If the scale shape is [1, 1, 5] and the input shape is [3, 4, 5], the ORT will throw an error. Or we inserted a reshape to support?

Reply:
> If the scale shape is [1, 1, 5] and the input shape is [3, 4, 5], the ORT will throw an error. Or we inserted a reshape to support?

Yes, I suppose we should squeeze (reshape) the scale tensor to 1D. And we need to find the axis from webnn scale tensor.